[SPARK-17629][ML] methods to return synonyms directly

Krimit opened the pull request:
## What changes were proposed in this pull request?
provide methods to return synonyms directly, without wrapping them in a dataframe

In performance sensitive applications (such as user facing apis) the roundtrip to and from dataframes is costly and unnecessary

The methods are named ``findSynonymsArray`` to make the return type clear, which also implies a local datastructure
## How was this patch tested?
updated word2vec tests

Commit:
provide methods to return synonyms directly, without wrapping them in a dataframe

In performance sensitive applications (such as user facing apis) the roundtrip to and from dataframes is costly and unnecessary

AmplabJenkins commented:
Can one of the admins verify this patch?

Krimit commented:
cc @jkbradley

srowen commented on mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala:
```diff
@@ -232,19 +232,40 @@ class Word2VecModel private[ml] (
+
+  /**
+   * Find "num" number of words closest in similarity to the given word, not
+   * including the word itself. Returns a dataframe with the words and the
```
(This doesn't return a DataFrame)

Krimit replied:
🔍

Commit:
update the javadoc

use @return to specify return values. Correct the return type for `findSynonymsLocal`

jkbradley commented:
Thanks for the PR!

What about findSynonymsArray?  That still implies a local value and is more specific.

Also, can you please add a unit test for this?

Commit:
rename to findSynonymsArray

is clearer both about the return type and that it runs locally
also - add a explicit test

Krimit commented:
@jkbradley - updated

Added an explicit new test as requested, although the existing test already covers it (by virtue of the existing methods calling the new methods)

felixcheung commented:
@srowen could you please kick Jenkins to test this PR? :)

srowen commented:
Jenkins add to whitelist

srowen commented:
Hm you should be able to restart it by asking here. If it doesn't it may be stuck or busy. You can also trigger at spark-prs.appspot.com

SparkQA commented:
**[Test build #72961 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72961/testReport)** for PR 16811 at commit [`7988385`](https://github.com/apache/spark/commit/7988385e8412a0176d1595f9d59dda843d0e4e23).

felixcheung commented:
AFAIK Jenkins won't listen to me or Yanbo. We tried couple times with other PRs - can try again next time.

SparkQA commented:
**[Test build #72961 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/72961/testReport)** for PR 16811 at commit [`7988385`](https://github.com/apache/spark/commit/7988385e8412a0176d1595f9d59dda843d0e4e23).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/72961/
Test PASSed.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala:
```diff
@@ -144,6 +144,31 @@ class Word2VecSuite extends SparkFunSuite with MLlibTestSparkContext with Defaul
     }
   }
 
+  test("findSynonymsArray") {
```
Can you please combine this with the findSynonyms test to avoid fitting the same model twice and to share code?

jkbradley replied:
Once you combine them, you can check findSynonymsArray by comparing it against findSynonyms.collect

jkbradley reviewed (commented):
Thanks for adding the test!

Also, please update the PR description since it will become part of the commit message.

Commit:
combine into one test

avoid training a new model. Instead just assert that both methods return the same results

Commit:
use existing fuzzyEquals implementation

rather than rolling a new implementation for each test

SparkQA commented:
**[Test build #73933 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73933/testReport)** for PR 16811 at commit [`2cca29a`](https://github.com/apache/spark/commit/2cca29a9b2c4e779ab6d3282d4025786868d9dbb).

Commit:
simplify case class

This is more readable. Use the nice features of scala!

Krimit commented:
Thanks for your comments @jkbradley, updated. I also took the opportunity to replace the kinda-janky fuzzyEquals in the test with the ``TestingUtils`` implementation

SparkQA commented:
**[Test build #73933 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73933/testReport)** for PR 16811 at commit [`2cca29a`](https://github.com/apache/spark/commit/2cca29a9b2c4e779ab6d3282d4025786868d9dbb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73933/
Test PASSed.

SparkQA commented:
**[Test build #73934 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/73934/testReport)** for PR 16811 at commit [`881353d`](https://github.com/apache/spark/commit/881353d7d104f3bff9246418c6149af6513c858b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/73934/
Test PASSed.

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/feature/Word2VecSuite.scala:
```diff
@@ -134,13 +134,20 @@ class Word2VecSuite extends SparkFunSuite with MLlibTestSparkContext with Defaul
+      case (expected, actual) => assert(expected ~== actual absTol 1E-5)
+    }
+
+    result.zip(model.findSynonymsArray("a", 2)).foreach {
```
Technically, collect() may not return elements in the same order each time, so this will be more robust if we convert each result to a map: synonym -> similarity and compare the maps.  Could you please do that?

Krimit replied:
Sure thing! I wonder if ``.collect()`` has ever caused flappy tests for this reason here or elsewhere

MLnick replied:
It has - usually in this case either a Map as suggested, or explicitly sorting the result of `collect` before comparison, is best.

jkbradley reviewed (commented):
Thanks for the updates!  Just one last comment

Commit:
use map for result assertions

results may not be in the same order

SparkQA commented:
**[Test build #74001 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74001/testReport)** for PR 16811 at commit [`3a02800`](https://github.com/apache/spark/commit/3a02800dbb3bf03f5ecb58f15691848d8df77cf0).

Krimit commented:
Updated. I do kind of wish we had access to ``assertJ``, which would make unordered assertions a cakewalk

SparkQA commented:
**[Test build #74001 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/74001/testReport)** for PR 16811 at commit [`3a02800`](https://github.com/apache/spark/commit/3a02800dbb3bf03f5ecb58f15691848d8df77cf0).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/74001/
Test PASSed.

jkbradley commented:
LGTM
Merging with master
Thank you!

jkbradley commented:
Thanks!  I made a follow-up JIRA for updating the Python API: https://issues.apache.org/jira/browse/SPARK-19866
